torch/_refs/__init__.py

```diff
@@ -2889,6 +2890,54 @@ def vsplit(
     return tensor_split(a, split_sizes, 0)
 
 
+@register_decomposition(torch.ops.aten.diagonal, disable_meta=True)
+def diagonal(
+    self: TensorLikeType,
+    offset: int = 0,
+    dim1: int = 0,
+    dim2: int = 1,
+) -> TensorLikeType:
+    """
+    Reference implementation of torch.diagonal
+    """
+    num_dims = self.dim()
+    dim1 = utils.canonicalize_dim(idx=dim1, rank=num_dims)
+    dim2 = utils.canonicalize_dim(idx=dim2, rank=num_dims)
+
+    check(
+        dim1 != dim2, lambda: f"diagonal dimensions cannot be identical {dim1}, {dim2}"
+    )
+
+    storage_offset = self.storage_offset()
+
+    if offset >= 0:
+        diag_size = max(min(self.size()[dim1], self.size()[dim2] - offset), 0)
+    else:
+        diag_size = max(min(self.size()[dim1] + offset, self.size()[dim2]), 0)
+
+    if diag_size == 0:
+        pass  # skip
+    elif offset >= 0:
+        storage_offset += offset * self.stride()[dim2]
```

Comment:
```suggestion
        storage_offset += offset * self.stride(dim2)
```

Reply:
No, please don't skip meta tests if you can avoid it (and it's wrong that there are other refs that are skipping meta, we need refs so that meta implementations work, skipping meta tests defeats the whole purpose).